servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherConcatWithCompletable.java

```diff
@@ -31,20 +31,154 @@
     private final Completable next;
+    private final boolean propagateCancel;
 
-    PublisherConcatWithCompletable(Publisher<T> original, Completable next) {
+    PublisherConcatWithCompletable(Publisher<T> original, Completable next, boolean propagateCancel) {
         super(original);
         this.next = requireNonNull(next);
+        this.propagateCancel = propagateCancel;
     }
 
     @Override
     public Subscriber<? super T> apply(Subscriber<? super T> subscriber) {
-        return new ConcatSubscriber<>(subscriber, next);
+        return propagateCancel ?
+                new ConcatSubscriberCancel<>(subscriber, next) : new ConcatSubscriber<>(subscriber, next);
+    }
+
+    private static final class ConcatSubscriberCancel<T>
+            implements CompletableSource.Subscriber, PublisherSource.Subscriber<T>, Subscription {
+        private static final Cancellable CANCELLED = () -> { };
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<ConcatSubscriberCancel, Cancellable> cancellableUpdater =
+                newUpdater(ConcatSubscriberCancel.class, Cancellable.class, "cancellable");
+        private final Subscriber<? super T> target;
+        private final Completable next;
+        private volatile Cancellable cancellable = IGNORE_CANCEL;
+
+        ConcatSubscriberCancel(Subscriber<? super T> target, Completable next) {
+            this.target = target;
+            this.next = next;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            cancellable = new FirstSubscription(s);
+            target.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T t) {
+            target.onNext(t);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            for (;;) {
+                final Cancellable c = cancellable;
+                assert c != IGNORE_CANCEL;
+                if (FirstSubscription.class.equals(c.getClass())) {
+                    if (cancellableUpdater.compareAndSet(this, c, CANCELLED)) {
+                        try {
+                            target.onError(t);
+                        } finally {
+                            next.subscribeInternal(this);
+                        }
+                        break;
+                    }
+                } else if (c == CANCELLED) {
+                    // CANCELLED means downstream doesn't care about being notified and internal state tracking also
+                    // uses this state if terminal has been propagated, so avoid duplicate terminal. We may subscribe
+                    // to both sources in parallel if cancellation occurs, so allowing terminal to propagate would mean
+                    // ordering and concurrency needs to be accounted for between Publisher and completed, because
+                    // cancel allows for no more future delivery we avoid future invocation of the target subscriber.
```

Comment:
Thanks for clarifying this. While we can ignore terminal events after cancel, we try to still deliver all originally emitted events in other operators. It's good if "deliver terminal after cancel" decision can be honored from the original source instead of modified by an operator.
IIUC, we need one more state to handle this. `cancel` needs to use CAS instead of getAndSet to flip subscription into that new state and all other states to `CANCELLED`.

Reply:
I will add a new state to propagate here.

Reply:
actually this fails the tck bcz we don't necessarily know if the publish was done emitting or not. lets not do this for now.